### PR TITLE
Enforce skip == true, if all tx blocks are zeros

### LINKED ISSUE #1319: Missing logic to detect all tx blocks are zero coeffs thus set skip flag = true

### PR DESCRIPTION
**Pull request by ycho**
Split from #1260
Resolves #1319

### CODE REVIEW

**Comment by ycho** [comment]
[awcy](https://beta.arewecompressedyet.com/?job=master-b074992e54ab4067b3eabecd45110d451da95c2d&job=enforce_skip_s6%402019-09-25T22%3A20%3A51.729Z)

At default speed, we have at least a small gain.

|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -0.1848 | -0.0391 |  0.4907 |  -0.1230 | -0.0953 |  0.0239 |    -0.0388 |

File: src/encoder.rs [hunk 1]
```diff
@@ -2075,6 +2084,9 @@ pub fn encode_block_with_modes<T: Pixel>(
   let mode_context =
     cw.find_mvrefs(tile_bo, ref_frames, &mut mv_stack, bsize, fi, is_compound);
 
+  if !mode_decision.skip && !mode_decision.has_coeff {
```
Comment (tdaede) [inline src/encoder.rs]: Does the !mode_decision.skip condition do anything?

Comment (ycho) [reply src/encoder.rs]: Yes, in mode_decision_rdo() we test two skip modes, true and false, and that is kept by  mode_decision.skip. After mode_decision is done, and if mode_decision.skip is false but mode_decision.has_coeff is false, this means the case that there is no non-zero coeff while skip is set false. This condition can be checked possibly at earlier stage, but I ended yup at this place.

Comment (ycho) [reply src/encoder.rs]: And I guess you probably thought that has_coeff is always false if mode_decision.skip is true. Yes, that is right.

**Comment by ycho** [comment]
> I like the concept, glad it gives gains.

Sure- thanks!

**Comment by ycho** [comment]
Little more gains with speed 0.
[awcy at speed 0](https://beta.arewecompressedyet.com/?job=master-b074992_s0%402019-09-25T22%3A18%3A03.180Z&job=enforce_skip_s0%402019-09-25T22%3A20%3A57.274Z)

|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -0.5058 | -1.4495 | -0.7322 |  -0.5344 | -0.6325 | -0.6209 |    -0.7418 |

**Pull request merged**